### PR TITLE
[Infra] Custom Hook to listen to async calls

### LINKED ISSUE #10: [BUG] should have generic infra to debounce/disable buttons while sending
image

potential way: https://blog.logrocket.com/rxjs-with-react-hooks-for-state-management/

### PR DESCRIPTION
**Pull request by wangtjwork**
Added @testing-library/react-hooks and react-test-renderer to test custom hooks.

Addressing https://github.com/tech-gamers/side-project-incubator-web/issues/10

### CODE REVIEW

File: src/app/hooks/useStatusForAsyncCall.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { useState } from 'react';
+
+export default <T>(call: () => Promise<T>): [boolean, () => Promise<T>] => {
+  const [status, setStatus] = useState(false);
+
+  const wrappedCall = async () => {
+    setStatus(true);
+
+    const callResult = await call();
+    setStatus(false);
```
Comment (yc-shawn) [inline src/app/hooks/useStatusForAsyncCall.ts]: Instead use true or false.
Maybe we can use constants, boolean is a little bit confused.

Initial: `IDLE` or `INIT`
after wrappedCall started: `IN_PROGRESS` or `BUSY`
after async call resolved: `COMPLETE`

Comment (yc-shawn) [reply src/app/hooks/useStatusForAsyncCall.ts]: OR

We still use boolean, but change `status` to `busy`.
init: `busy = fasle`,
in progress: `busy = true`,
completed: `busy = false`

Comment (wangtjwork) [reply src/app/hooks/useStatusForAsyncCall.ts]: Using constant enum.

File: src/app/hooks/useStatusForAsyncCall.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+  const wrappedCall = async () => {
+    setStatus(new StatusState(Status.BUSY));
+
+    const callResult = await call();
+    setStatus(new StatusState(Status.COMPLETED));
+    return callResult;
+  };
+  return [status, wrappedCall];
+};
```
Comment (robturtle) [inline src/app/hooks/useStatusForAsyncCall.ts]:
```suggestion
export type AsyncStatus = "idle" | "busy" | "completed" | "error";

export default <T>(call: () => Promise<T>): [AsyncStatus, () => Promise<T>] => {
  const [status, setStatus] = useState("idle");

  async function wrappedCall() => {
    try {
      setStatus("busy");
      const result = await call();
      setStatus("completed");
      return result;
    } catch(error) {
      setStatus("error");
      throw error;
    }
  }
  
  return [status, wrappedCall];
}
```

Comment (robturtle) [reply src/app/hooks/useStatusForAsyncCall.ts]: 顺便建议改名为从属短语：`useAsyncCallStatus`

**Pull request merged**